src/processor/operator/recursive_extend/bfs_state.cpp

```diff
@@ -0,0 +1,203 @@
+#if defined(_MSC_VER)
+#include <Windows.h>
+#include <intrin.h>
+#endif
+#include "common/exception.h"
+#include "processor/operator/recursive_extend/bfs_state.h"
+
+namespace kuzu {
+namespace processor {
+
+void SSSPSharedState::reset(TargetDstNodes* targetDstNodes) {
+    ssspLocalState = EXTEND_IN_PROGRESS;
+    currentLevel = 0u;
+    nextScanStartIdx = 0u;
+    numVisitedNodes = 0u;
+    auto totalDestinations = targetDstNodes->getNumNodes();
+    if (totalDestinations == (maxOffset + 1) || totalDestinations == 0u) {
+        // All node offsets are destinations hence mark all as not visited destinations.
+        std::fill(visitedNodes.begin(), visitedNodes.end(), NOT_VISITED_DST);
+    } else {
+        std::fill(visitedNodes.begin(), visitedNodes.end(), NOT_VISITED);
+        for (auto& dstOffset : targetDstNodes->getNodeIDs()) {
+            visitedNodes[dstOffset.offset] = NOT_VISITED_DST;
+        }
+    }
+    std::fill(pathLength.begin(), pathLength.end(), 0u);
+    bfsLevelNodeOffsets.clear();
+    srcOffset = 0u;
+    numThreadsActiveOnMorsel = 0u;
+    nextDstScanStartIdx = 0u;
+    inputFTableTupleIdx = 0u;
+}
+
+/*
+ * Returning the state here because if SSSPSharedState is complete / in pathLength writing stage
+ * then depending on state we need to take next step. If MORSEL_COMPLETE then proceed to get a new
+ * SSSPSharedState & if MORSEL_pathLength_WRITING_IN_PROGRESS then help in this task.
+ */
+bool SSSPSharedState::getBFSMorsel(
+    std::unique_ptr<BaseBFSMorsel>& bfsMorsel, SSSPLocalState& state) {
+    std::unique_lock<std::mutex> lck{mutex};
+    switch (ssspLocalState) {
+    case MORSEL_COMPLETE: {
+        state = NO_WORK_TO_SHARE;
+        return true;
+    }
+    case PATH_LENGTH_WRITE_IN_PROGRESS: {
+        bfsMorsel->ssspSharedState = this;
+        state = ssspLocalState;
+        return true;
+    }
+    case EXTEND_IN_PROGRESS: {
+        if (nextScanStartIdx < bfsLevelNodeOffsets.size()) {
+            numThreadsActiveOnMorsel++;
+            auto bfsMorselSize = std::min(
+                common::DEFAULT_VECTOR_CAPACITY, bfsLevelNodeOffsets.size() - nextScanStartIdx);
+            auto morselScanEndIdx = nextScanStartIdx + bfsMorselSize;
+            auto shortestPathMorsel =
+                (reinterpret_cast<ShortestPathMorsel<false>*>(bfsMorsel.get()));
+            shortestPathMorsel->reset(nextScanStartIdx, morselScanEndIdx, this);
+            nextScanStartIdx += bfsMorselSize;
+            return false;
+        }
+        state = NO_WORK_TO_SHARE;
+        return true;
+    }
+    default:
+        throw common::RuntimeException(
+            &"Unknown local state encountered inside SSSPSharedState: "[ssspLocalState]);
+    }
+}
+
+bool SSSPSharedState::hasWork() const {
+    if (ssspLocalState == EXTEND_IN_PROGRESS && nextScanStartIdx < bfsLevelNodeOffsets.size()) {
```

Comment:
It's recommended to do 
```
if 
    return
if 
    return
```

rather than
```
if 
   return
else if
   return
```
Apply across PR if you want

Reply:
changed